### PR TITLE
Refactor and Add Tests

### LINKED ISSUE #5: Add tests
Tests should be added to the `runtests.jl` file. Ideally some way of emulating an FT device should be found. Otherwise this can check that errors are thrown as expected in the absence of any device.

### LINKED ISSUE #8: Split high and low level wrappers
Basic IO functionality could possibly be split from the C function wrappers.

### LINKED ISSUE #9: Add Documenter comments to all functions
Documentation should be added to all functions to enable help support.

### LINKED ISSUE #16: Finaliser often fails
The garbage collector's running of the finaliser often fails giving an error 

```
error in running finalizer: LibFTD2XX.FT_STATUS_ENUM(0x00000001)
```

### PR DESCRIPTION
**Pull request by ReubenHill**
Big pull request which changes the API and makes everything much more robust.

## Notes
 
 - Only exports what is necessary
 - Splits front and back end. Back end is now accessed via submodule `LibFTD2XX.Wrapper` (resolves #8)
 - Supports much more back-end functionality
 - Massively improves test coverage to cover all functions (resolves #5) 
 - Much improved documentation (resolves #9)
 - resolves #16 and generally makes everything much more robust
 - drops 0.6 support
 - update licence/copyright notices

## Issues
 - Only cross-platform (OSX, Linux and Windows) features supported. For example there is no support yet of `FTOpenEx` with `FT_OPEN_BY_LOCATION` flag.
 - Some cross-platform library functions are not yet supported.
 - Needs auto-generated documentation.
 - `.Wrapper` functions which fail throw `FT_STATUS_ENUM`s instead of exceptions. This should be replaced with a suitable exception type.
 - timeouts = 0 are allowed but their behaviour is currently undefined. This needs investigation.
 - test on linux and macos with hardware
 - CI fails on linux but works on local linux machines. (See report at https://travis-ci.org/Gowerlabs/LibFTD2XX.jl/builds/511412464)

## Todo
 - [x] Document changes
 - [x] Document all existing issues
   - [x] Only cross-platform (OSX, Linux and Windows) features supported. For example there is no support yet of `FTOpenEx` with `FT_OPEN_BY_LOCATION` flag.
   - [x] Missing cross-platform functions in library
   - [x] Needs auto-generated documentation
 - [x] Test on Julia 0.6
 - [x] Support timeout setting on front end
 - [x] Update readme.md
 - [x] Test with real use case to make sure nothing is broken
 - [x] Stop functions that are not macos/linux compatible running on those platforms
 - [x] Deal with CI failure on linux

### CODE REVIEW

**Comment by samuelpowell** [comment]
Drop 0.6 for sure

**Comment by ReubenHill** [comment]
Ready for review. @samuelpowell do you know how to get codecoverage to update?

**Comment by samuelpowell** [comment]
@ReubenHill it may be that the configuration scripts are out of date. I would suggest using `PkgTemplates.jl` to create the files you want with the appropriate options and then graft them into this repo as appropriate.

File: src/util.jl
Comment (samuelpowell) [inline src/util.jl]: Is it nor normal to write `ntuple2string(input) where {N}`, e.g the `where` is outside the brackets, check style guide

File: src/util.jl
Comment (samuelpowell) [inline src/util.jl]: Compat required for what? 0.6 dropped now, won't flag again - find all instances, and remove Compat dependency if no longer required.

File: src/util.jl
Comment (samuelpowell) [inline src/util.jl]: You know what the length is, it is `N`

File: src/util.jl
Comment (samuelpowell) [inline src/util.jl]: Is this a method error? Looks like an `ArgumentError`

File: src/util.jl
Comment (samuelpowell) [inline src/util.jl:36]: This (and other things) is potentially very poor for performance - is this function called a lot?

File: test/runtests.jl
Comment (samuelpowell) [inline test/runtests.jl:13]: Run no hardware tests anyway?

File: test/nohardware/wrapper.jl
Comment (samuelpowell) [inline test/nohardware/wrapper.jl]: Be careful - is this true on all platforms? I think I recall it returning different things on Linux/Mac, indeed the version is different across platforms

File: src/wrapper.jl
Comment (samuelpowell) [inline src/wrapper.jl]: Should this be exported? Quite likely candidate for user variable names/other packages

File: src/wrapper.jl
Comment (samuelpowell) [inline src/wrapper.jl]: Whitespace <:

File: src/wrapper.jl
Comment (samuelpowell) [inline src/wrapper.jl]: These are really dirty, prefix as `_ptr(....)` to really emphasise for internal use!?

File: src/wrapper.jl
Comment (samuelpowell) [inline src/wrapper.jl]: These comments aren't useful - you're just repeating the function name, similarly no need to refer to the programmers guide - put this all in the README and remove here and below to reduce LOC and noise on ?help

File: src/wrapper.jl
Comment (samuelpowell) [inline src/wrapper.jl:357]: Why is this a tuple of characters anyway?

File: src/wrapper.jl
Comment (samuelpowell) [inline src/wrapper.jl:233]: Why is this a tuple of Chars, seems very strange

File: src/wrapper.jl
Comment (samuelpowell) [inline src/wrapper.jl:234]: As above

File: src/wrapper.jl
Comment (samuelpowell) [inline src/wrapper.jl:366]: No throw?

File: src/wrapper.jl
Comment (samuelpowell) [inline src/wrapper.jl]: This line is repeated everywhere make a function that takes status e..g `_checkstatus(status) = ....`

File: src/wrapper.jl
Comment (samuelpowell) [inline src/wrapper.jl]: What exactly are you throwing here and elsewhere?

**Review by samuelpowell — COMMENTED** [review]
This looks a lot better than before, thank you! Initial comments...

**Comment by samuelpowell** [comment]
@ReubenHill will clone and look through in further detail once you've addressed above

**Comment by samuelpowell** [comment]
Also, I know it's a bit late now, but had you not considered using `Clang.jl`? See e.g. `Spinnaker.jl`to see what it does.

File: src/util.jl
Comment (ReubenHill) [reply src/util.jl:36]: No, never. It's there for convenience if anyone every wants to use `FT_GetDeviceInfoList` which is only accessible to someone using the `.Wrapper` submodule. The high-level stuff in LibFTD2XX.jl uses the much nicer `FT_GetDeviceInfoDetail` to get details about devices now.

File: test/nohardware/wrapper.jl
Comment (ReubenHill) [reply test/nohardware/wrapper.jl]: The docs were not clear on this point. I'm tempted to leave it and then find out if this fails on Linux etc and fix at that point. Thoughts?

File: test/runtests.jl
Comment (ReubenHill) [reply test/runtests.jl:13]: The "no hardware" tests currently _require_ that there be FTD2XX no hardware present. They fail if it's present.

**Comment by ReubenHill** [comment]
> Also, I know it's a bit late now, but had you not considered using Clang.jl? See e.g. Spinnaker.jlto see what it does.

I decided learning how to use Clang.jl would take more time than expanding what I had already to cover more functions. In hindsight, that might have been wrong. The majority of this has been adding tests, making more robust and and adding high level stuff.

File: src/wrapper.jl
Comment (ReubenHill) [reply src/wrapper.jl]: This is not a case of[ type piracy](https://docs.julialang.org/en/v1/manual/style-guide/#Avoid-type-piracy-1) - what's the issue if it's exported? The `Wrapper` submodule is only accessible as a submodule of `LibFTD2XX` anyway. Other people would have to explicitly import it to screw their namespaces

File: src/util.jl
Comment (ReubenHill) [reply src/util.jl]: fixed in 913629eab9961a4149e5b7ae84baf2b54d9fcd2f

File: src/util.jl
Comment (ReubenHill) [reply src/util.jl]: Resolved in 76caea9553a6f76f7d912b5b5376938c439eceb7

File: src/wrapper.jl
Comment (ReubenHill) [reply src/wrapper.jl]: resolved in e632c86bf430902c45b7c313c54c6f7ae40505b7

File: src/util.jl
Comment (ReubenHill) [reply src/util.jl]: resolved in 216d743a5996f060b3907cd8da19295d1be4fcc6

File: src/util.jl
Comment (ReubenHill) [reply src/util.jl]: resolved in 216d743a5996f060b3907cd8da19295d1be4fcc6

File: src/wrapper.jl
Comment (ReubenHill) [reply src/wrapper.jl]: resolved in 72caa5437d75e79248ff099d7f8366b73db88311

File: src/wrapper.jl
Comment (ReubenHill) [reply src/wrapper.jl]: resolved in e632c86bf430902c45b7c313c54c6f7ae40505b7

File: src/wrapper.jl
Comment (ReubenHill) [reply src/wrapper.jl]: Resolved in 1203b37a22480876b03880e174f2549a070c47f1 and 02bdaf3ab5d5f71b75753eb65fda1962f910cd1f

File: src/wrapper.jl
Comment (ReubenHill) [reply src/wrapper.jl:233]: This is a struct which is passed directly into `FT_GetDeviceInfoList`. The `serialnumber` and `description` within that struct should be pre-allocated `char*` arrays. See the documentation on [struct type correspondences](https://docs.julialang.org/en/v1/manual/calling-c-and-fortran-code/#Struct-Type-correspondences-1).

Comment (ReubenHill) [reply src/wrapper.jl:233]: I've added a note in 17a99f24f4996743880a1f61c4eb5ef5d87aa32e

File: src/wrapper.jl
Comment (ReubenHill) [reply src/wrapper.jl:357]: See other comment

File: src/wrapper.jl
Comment (ReubenHill) [reply src/wrapper.jl:234]: See other comment

File: src/wrapper.jl
Comment (ReubenHill) [reply src/wrapper.jl:366]: Good spot! Fixed in dac0d67c7c433682b99b2eb0f1d0d59c3c45334c

File: src/wrapper.jl
Comment (ReubenHill) [reply src/wrapper.jl]: An `FT_STATUS_ENUM`. It should actually be a special exception type. This is noted as a problem in issues in main PR comment:

>  - `.Wrapper` functions which fail throw `FT_STATUS_ENUM`s instead of exceptions. This should be replaced with a suitable exception type.

File: src/wrapper.jl
Comment (ReubenHill) [reply src/wrapper.jl]: resolved in 3a7e05f49eca09bf4edc8c1d05e19eb82e93ffaa

File: src/wrapper.jl
Comment (samuelpowell) [reply src/wrapper.jl:233]: So the C side of this is a `char bla[N]`?

File: src/wrapper.jl
Comment (samuelpowell) [reply src/wrapper.jl]: Open issue?

File: src/util.jl
Comment (samuelpowell) [reply src/util.jl:36]: Can you not use unsafe_string(pointer(input[1]), endidx) here?

Comment (samuelpowell) [reply src/util.jl:36]: No of course this is stack allocated, disregard

Comment (samuelpowell) [reply src/util.jl:36]: `String(collect(char[1:end]))`?

Comment (samuelpowell) [reply src/util.jl:36]: No, leave it! It's fine

File: src/LibFTD2XX.jl
Comment (samuelpowell) [inline src/LibFTD2XX.jl]: Just splat it

```
D2XXDevice(getdeviceinfodetail(deviceidx)...)
```

File: src/LibFTD2XX.jl
Comment (samuelpowell) [inline src/LibFTD2XX.jl:309]: Might be worth commenting that the vector will be resized if not large enough

Comment (samuelpowell) [reply src/LibFTD2XX.jl:309]: Also, if you resize an array I assume that it might change memory address. It will still be correctly bound (and thus still mutating the same object) right?

**Review by samuelpowell — APPROVED** [review]
There are a couple of comments below but this LGTM assuming CI passes, and that it works properly - I can't check locally at the moment. I think the simplicity of the Julia side interface demonstrates this is well thought out!

File: src/wrapper.jl
Comment (ReubenHill) [reply src/wrapper.jl:233]:
```julia
struct FT_DEVICE_LIST_INFO_NODE
   ...
   serialnumber::NTuple{16, Cchar}
   ...
end
```
is 
```C
typedef struct _ft_device_list_info_node {
   ...
   char SerialNumber[16];
   ...
} FT_DEVICE_LIST_INFO_NODE
```

File: src/wrapper.jl
Comment (ReubenHill) [reply src/wrapper.jl]: I will open issues for everything in the issue list of the PR once this is approved

File: src/LibFTD2XX.jl
Comment (ReubenHill) [reply src/LibFTD2XX.jl:309]: > Might be worth commenting that the vector will be resized if not large enough

This is the standard behaviour for `readbytes!` for an `<: IO` and is already documented. From the standard repl:
```julia
julia> using LibFTD2XX
[ Info: Recompiling stale cache file C:\Users\Reuben\.julia\compiled\v1.0\LibFTD2XX\U18mP.ji for LibFTD2XX [505cea47-0b8c-5d8e-a4f7-f3ac0095ebab]

help?> readbytes!
search: readbytes!

  readbytes!(stream::IO, b::AbstractVector{UInt8}, nb=length(b))

  Read at most nb bytes from stream into b, returning the number of bytes
  read. The size of b will be increased if needed (i.e. if nb is greater than
  length(b) and enough bytes could be read), but it will never be decreased.

  ────────────────────────────────────────────────────────────────────────────

  readbytes!(stream::IOStream, b::AbstractVector{UInt8}, nb=length(b); all::Bool=true)

  Read at most nb bytes from stream into b, returning the number of bytes
  read. The size of b will be increased if needed (i.e. if nb is greater than
  length(b) and enough bytes could be read), but it will never be decreased.

  See read for a description of the all option.

  ────────────────────────────────────────────────────────────────────────────

  readbytes!(d::D2XXDevice, b::AbstractVector{UInt8}, nb=length(b))

  See also: D2XXDevice.

  ────────────────────────────────────────────────────────────────────────────

  readbytes!(handle::FT_HANDLE, b::AbstractVector{UInt8}, nb=length(b))

  See also: FT_HANDLE
```
Is this sufficiently non-ambiguous? Seems pointless to add an identical comment to this method


> Also, if you resize an array I assume that it might change memory address. It will still be correctly bound (and thus still mutating the same object) right?

 I don't follow?

File: src/LibFTD2XX.jl
Comment (ReubenHill) [reply src/LibFTD2XX.jl]: resolved in b701a0b9fe82bdab53bf0c02a774d4dc99d3a382

File: src/LibFTD2XX.jl
Comment (ReubenHill) [reply src/LibFTD2XX.jl:309]: I've updated the `readbytes!` comment in d61c7d01d9241bebd9544ac580943855fe9de38d

**Pull request merged**
